### ISSUE TITLE
Type hints

### ISSUE THREAD

**Issue by adamchainz** [wolever/parameterized#143]
I've been rolling out Mypy on a client project. When activating the [`disallow_untyped_decorators` flag](https://mypy.readthedocs.io/en/stable/config_file.html#confval-disallow_untyped_decorators), I found I needed type hints for parametrized. I added the below to the project's stubs folder (on [`mypy_path`](https://mypy.readthedocs.io/en/stable/config_file.html#confval-mypy_path):

```python
from __future__ import annotations

from collections.abc import Callable, Iterable, Sequence
from typing import Any, ParamSpec, TypeVar
from unittest import TestCase

P = ParamSpec("P")
T = TypeVar("T")

class param:
    def __init__(self, *args: Any, **kwargs: Any) -> None: ...

InputType = (
    Iterable[str] | Iterable[Sequence[Any]] | Iterable[dict[str, Any]] | Iterable[param]
)

class parameterized:
    def __init__(
        self,
        input: InputType,
        doc_func: Callable[[Callable[..., Any], int, param], str] | None = None,
        skip_on_empty: bool = False,
    ): ...
    def __call__(self, test_func: Callable[P, T]) -> Callable[P, T]: ...
    @classmethod
    def expand(
        cls,
        input: InputType,
        name_func: Callable[[Callable[..., Any], int, param], str] | None = None,
        skip_on_empty: bool = False,
        **legacy: Any,
    ) -> Callable[[Callable[P, T]], Callable[P, T]]: ...

_TestCaseClass = TypeVar("_TestCaseClass", bound=type[TestCase])

def parameterized_class(input: Any) -> Callable[[_TestCaseClass], _TestCaseClass]: ...
```

These hints cover the main usage of parameterized. If you're interested, I wouldn't mind contributing full type hint coverage to the package.

**Comment by kasittig** [comment]
@adamchainz - an alternate place to contribute these stubs is the [typeshed project](https://github.com/python/typeshed)! I found your issue while enabling the mypy --strict flag on my own project.

**Comment by adamchainz** [comment]
Yeah that might be the solution given no response here. If you want to take my hints and open a typeshed PR please feel free!

**Comment by wolever** [comment]
Hey folks! Sorry for the delay here. I'd love to have type hint support, and would be more than happy to help get a PR working (although I confess I'm just coming back to Python after a couple of years in TypeScript land, and not yet fully up to speed on best practices).

If you could open a PR that gives an example of how to test that the project is covered by types, I'd be happy to help get them put in place.

**Comment by adamchainz** [comment]
I just had a look at doing this but found the python 2/3 compatibility code a bit "in the way". It will be hard to get Mypy to properly check it. I think all the compat code should be removed first.

**Comment by wolever** [comment]
Good news, @adamchainz! Python 2 support is dropped in 0.9! I haven't removed all the compat code yet, but if you can point me in the right direction for how to test with mypy, I'm happy to help iterate on it.

I'm going to cut a 0.9 release today, and adding typing support seems like a feature worthy of a 1.0 release.

**Comment by stephew2** [comment]
Any updates on v1.0 and/or typing mypy support for this?  Would love to have it!  Thanks.

**Comment by adamchainz** [comment]
I found adding type hints too hard so I wrote my own more minimal parametrised unittest library: https://pypi.org/project/unittest-parametrize/ . It copies pytest’s API as much as possible and comes with full types.
